### PR TITLE
add controls to order Hotcues by position

### PR DESCRIPTION
**Pull request by ronso0**
This is a small helper for my hotcue workflow, see below.

**TL;DR:**
Adds two controls to order hotcues by position in the track:
6 4 5 2 ==> 2 4 5 6
(only index is changed, see static hotcue colors in the waveform)
* `sort_hotcues` just reorder (keeps unset cues)
hotcues-sort-75

* `sort_hotcues_remove_offsets` reorder and compress: hotcues start at 1, remove unset cues
hotcues-sort-compress-75

### TODO
- [x] add tests
- [x] include loop cues and jump cues
- [x] move actions to 'Hotcue' menu (track menu)

There's no a 'Hotcues' menu with these two actions:
hotcues-menu

**Background:**
I usually set 1 at the start/first sound, 2 at the first beat (ifi t's not 1 already), 3 left empty for some interesting spot in long intros, 4 first phrase for mixing (main) beat etc.
So sometimes it turns out I don't need 2 & 3 but need some extra cues in between 4 - 8.
Or I need extra cues before 4.
Being able to rearrange cues with #13394 is great, but might be cumbersome, especially when I'm in a hurry during (live) preparation

Draft until #13394 has been merged and there's some agreement how to proceed here.

### CODE REVIEW

**Comment by Eve00000** [comment]
Thank you @ronso0 , I realy like this! Both thumbs up!
I also have my own numbering of hotcues, quite similar as yours.

File: src/track/track.cpp [hunk 1]
```diff
@@ -989,6 +989,63 @@ void Track::shiftCuePositionsMillis(double milliseconds) {
     markDirtyAndUnlock(&locked);
 }
 
+void Track::setHotcueIndicesSortedByPosition(bool compress) {
+    auto locked = lockMutex(&m_qMutex);
```
Comment (Swiftb0y) [inline src/track/track.cpp:993]: This would be a good candidate for a unit test. wdyt?

File: src/widget/wtrackmenu.cpp [hunk 2]
```diff
@@ -463,6 +463,23 @@ void WTrackMenu::createActions() {
...
+        connect(m_pSortHotcuesByPositionCompressAction,
+                &QAction::triggered,
+                this,
+                [this]() {
+                    slotSortHotcuesByPosition();
+                });
+        m_pSortHotcuesByPositionAction = new QAction(
+                tr("Sort hotcues by position (keep offsets)"), m_pClearMetadataMenu);
+        connect(m_pSortHotcuesByPositionAction,
+                &QAction::triggered,
```
[the remark this answers is not in the document]
Comment (ronso0) [reply src/widget/wtrackmenu.cpp:478]: Yeah, just "(remove offsets)" for the first case would be more clear.

File: src/track/track.cpp (hunk 1 above)
Comment (ronso0) [reply src/track/track.cpp:993]: :+1:

Comment (Swiftb0y) [reply src/track/track.cpp:993]: Do you have time to implement one?

Comment (ronso0) [reply src/track/track.cpp:993]: Sure, done.

**Comment by ywwg** [comment]
oh I would love to merge this in

**Comment by ywwg** [comment]
personally I assign hotcue numbers based on proportional position in the track.  1 is near the beginning of the track, 4 is a little past the middle, 8 is at the end.  This isn't always possible but helps keep them sorted manually

File: src/track/track.cpp [hunk 3]
```diff
@@ -989,6 +989,63 @@ void Track::shiftCuePositionsMillis(double milliseconds) {
     markDirtyAndUnlock(&locked);
 }
 
+void Track::setHotcueIndicesSortedByPosition(bool compress) {
```
Comment (ywwg) [inline src/track/track.cpp:992]: Even when there are only two options, I would prefer to make this an enum, not a bool.  This makes it more clear at the call site what the parameter is for.  (Also I would like to add my own scheme, assigning numbers by relative track position, which would be a third option here)

Comment (ywwg) [reply src/track/track.cpp:992]: that would be a post-merge change

Comment (Swiftb0y) [reply src/track/track.cpp:992]: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#i4-make-interfaces-precisely-and-strongly-typed

**Comment by ronso0** [comment]
Besides the enum aspect, how do you think we move forward here?
Just merge the COs, then discuss how to make them accessible in the GUI?
* track menu? it's huge aready and these options may not be needed
* two buttons in the cue menu?

**Comment by Swiftb0y** [comment]
> Just merge the COs, then discuss how to make them accessible in the GUI?

I'm not sure. We currently only have GUI language to talk about a single hotcue or the entire track... Adding a button that affects all hotcues to a menu that only discerns a single hotcue does not make sense to me. We could add another entry to the context menu or we could add some sort of visible container of the hotcues in the skins for it. There are plenty of (not yet implemented) actions that apply to all hotcues. Maybe some sort of extra cogwheel button next to the hotcue grid could work?

File: src/track/track.cpp (hunk 3 above)
Comment (ronso0) [reply src/track/track.cpp]: I made it an enum class and rebased.
Let me know when you want to start reviewing, I'll stop force-pushing then.

**Comment by ywwg** [comment]
Does everything need to be accessible from a controller? This is a very "preparing tracks" operation and does not seem like something that needs to be available at performance time.  Putting it only in the right-click menu seems fine to me

**Comment by ronso0** [comment]
No of course not. Here I have unconsciously started with the CO, then added the track menu action.

I noticed that already in the first session I used it a lot, I think on S4 I'll map it to Shift+Hotcues longpress or something (the padmode button) instead of via track menu. Now the actions are at the top level for quick access (demo) but they should indeed be in a Hotcue submenu.

**Comment by ronso0** [comment]
> We currently only have GUI language to talk about a single hotcue or the entire track... Adding a button that affects all hotcues to a menu that only discerns a single hotcue does not make sense to me. We could add another entry to the context menu or we could add some sort of visible container of the hotcues in the skins for it. There are plenty of (not yet implemented) actions that apply to all hotcues. Maybe some sort of extra cogwheel button next to the hotcue grid could work?

True, a dedicated 'all hotcues' edit section could be helpful.
Though I'd say we're good atm with the Clear action in the track menu, the shift controls next to the waveforms and color settings in the preferences.

**Comment by Swiftb0y** [comment]
sure... though in the long-term collecting those actions in their own menu would definitely make sense.

**Comment by ronso0** [comment]
Well, how to proceed here, what do you think?
Do you agree to move the actions into new 'Hotcue' menu now (track menu)?

**Comment by Swiftb0y** [comment]
yeah, lets do that. anything else is for a follup up PR.

**Comment by ronso0** [comment]
Okay, they're in the menu now.
Most importantly I included LoopCues and JumpCues (still unused but there's #13216) which would otherwise have been overwritten :grimacing: 

I decided to rebase to resolve the conflicts since all review comments have been addressed, hope that's okay?

File: src/widget/wtrackmenu.cpp [hunk 4]
```diff
@@ -2170,6 +2210,28 @@ void WTrackMenu::slotClearHotCues() {
...
+                "Sorting hotcues of %n track(s) by position (remove offsets)",
+                "",
+                getTrackCount());
+        break;
+    case HotcueSortMode::KeepOffsets:
+    default:
+        progressLabelText =
+                tr("Sorting hotcues of %n track(s) by position", "", getTrackCount());
+        break;
+    }
```
Comment (Swiftb0y) [inline src/widget/wtrackmenu.cpp:2227]: lets avoid the `default`. That we get warnings in case you add another sort mode and don't update the switch-case here.
```suggestion
    case HotcueSortMode::KeepOffsets:
        progressLabelText =
                tr("Sorting hotcues of %n track(s) by position", "", getTrackCount());
        break;
    }
```

**Review by Swiftb0y — COMMENTED** [review]
last nit

File: src/widget/wtrackmenu.cpp (hunk 4 above)
Comment (ronso0) [reply src/widget/wtrackmenu.cpp:2227]: I guess you mean to separate the `default` branch and print a warning, or even add a DEBUG_ASSERT there?

Comment (Swiftb0y) [reply src/widget/wtrackmenu.cpp]: No, just remove it completely. See this [godbolt](https://compiler-explorer.com/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:1,endLineNumber:29,positionColumn:1,positionLineNumber:29,selectionStartColumn:1,selectionStartLineNumber:29,startColumn:1,startLineNumber:29),source:'enum+class+E+%7B%0A++++A,%0A++++B,%0A%7D%3B%0A%0Aint+foo(E+v)+%7B%0A++++switch+(v)+%7B%0A++++//+commenting+out+one+of+the+below+cases+will+result+in+a+compile+error+(as+intended)%0A++++case+E::A:%0A++++++++return+42%3B%0A++++case+E::B:%0A++++++++return+43%3B%0A++++%7D%0A++++//+the+return+afterwards+is+needed+because+the+technical+definition+of+enums+is+weird+(to+allow+for+bitflags)+%0A++++return+0%3B%0A%7D%0A%0A%0Aint+bar(E+v)+%7B%0A++++switch+(v)+%7B%0A++++//+while+the+same+doesn!'t+happen+when+a+%60default%60+case+is+present:%0A++++//+case+E::A:%0A++++//+++++return+42%3B%0A++++default:%0A++++case+E::B:%0A++++++++return+43%3B%0A++++%7D%0A%7D%3B%0A'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:g142,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-Wall+-Werror',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+gcc+14.2+(Editor+%231)',t:'0')),k:50,l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+14.2',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+14.2+(Compiler+%231)',t:'0')),header:(),l:'4',m:50,n:'0',o:'',s:0,t:'0')),k:50,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) for a demonstration.

Comment (ronso0) [reply src/widget/wtrackmenu.cpp]: Okay, so you say the compiler warning is sufficient/better (as long as they're fatal)

Comment (Swiftb0y) [reply src/widget/wtrackmenu.cpp]: yes, much better than no warning and a best guess runtime handling. The more we can do/know at compile-time the better.

**Review by Swiftb0y — APPROVED** [review]
LGTM. Thank you. feel free to squash

**Comment by ywwg** [comment]
happy to get this in!

**Comment by Swiftb0y** [comment]
Thanks for the reminder
